rmf_fleet_adapter/include/rmf_fleet_adapter/agv/RobotUpdateHandle.hpp

```diff
@@ -110,6 +111,36 @@ class RobotUpdateHandle
   /// value that was given to the setter.
   rmf_utils::optional<rmf_traffic::Duration> maximum_delay() const;
 
+  /// Signature for a callback to be executed to let the fleet adapter know that
+  /// the robot has completed the requested action
+  using ActionCompleted = std::function<void()>;
```

Comment:
I see that things like `ActionCompleted` and `update_action_remaining_time` are separated right now even though they're conceptually coupled together and both coupled with the execution of a specific action instance. Instead of having them decoupled in the RobotUpdateHandle interface, how about this:

```
class ActionExecution
{
public:
  // Update the amount of time remaining for this action
  void update_remaining_time(rmf_traffic::Duration remaining_time_estimate);

  // Trigger this when the action is finished
  void finished();

  // Allow the action performer to report directly to the schedule.
  // This returns nullptr after `finished()` has been triggered.
  Participant* schedule();
  const Participant* schedule() const;

  // TODO: Consider giving access to the negotiation? Maybe in a future version

  ~ActionExecution()
  {
    // Automatically trigger finished when this object is lost
    finished();
  }

  class Implementation;
private:
  rmf_utils::unique_impl_ptr<Implementation> _pimpl;
};

using ActionExecutor = std::function<void(
  const std::string& category,
  const nlohmann::json& description,
  ActionExecution execution)>;
```

This pattern will allow us to extend the action execution API in the future (e.g. giving action executors the ability to negotiate traffic) while remaining backwards compatible.

Reply:
Implemented as discussed.